sbin/chkconfig/chkconfig.8
.\" Copyright 1993, Garrett A. Wollman.
.\" Copyright 1993, University of Vermont and State Agricultural College.
.\" All Rights Reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE UNIVERSITY AND AUTHOR ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE UNIVERSITY OR AUTHOR BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $Id: chkconfig.8,v 1.1 1993/11/11 05:25:28 wollman Exp $
.\"
.Dd November 10, 1993
.Dt CHKCONFIG 8
.Os "FreeBSD"
.Sh NAME
.Nm chkconfig
.Nd "set and inquire the state of system configuration files"
.Sh SYNOPSIS
.Nm chkconfig
.Op Fl d Ar dirname
.Op Fl s
.Nm chkconfig
.Op Fl d Ar dirname
.Op Fl v
.Ar option
.Nm chkconfig
.Op Fl d Ar dirname
.Op Fl f
.Op Fl v
.Ar option
.Ar state
.Sh DESCRIPTION
The
.Nm chkconfig
program is used to set and inquire the state of the system
configuration files.  These files are used by the system startup
scripts
.Pa /etc/rc ,
.Pa /etc/netstart ,
and
.Pa /etc/rc.local
to determine which programs and services to start when the system
boots.  The configuration files are normally located in the directory
.Pa /etc/config ,
but an alternate location can be specified with the
.Fl d
flag.
By convention, configuration file names do not contain dots (which are
reserved for files containing command-line options for various
programs), but this is not strictly enforced.
.Pp
There are three forms of the
.Nm chkconfig
command.  The first form, with no
.Ar option
argument, lists the state of the entire set of options.  If the
.Fl s
flag is specified, the files are first sorted according to the state
of each option; by default, they are sorted in ASCII collation order.
.Pp
The second form, with only a
.Ar option
argument (and no state) specified, retrieves the current value of the
specified option.  If the appropriate file exists and contains only
the word ``on'', then the option is considered to be on; otherwise it
is considered to be off.  If the
.Fl v
flag is specified,
.Nm chkconfig
displays the status on the standard output.  In either case,
.Nm chkconfig
exits with a status of zero (true) if the option is on, or 1 (false)
otherwise.
.Pp
The third form, with both
.Ar option
and
.Ar state
specified, attempts to change the state of
.Ar option
to be
.Ar state ,
which is interpreted in the same way as the file contents (above).  If
the configuration file does not already exist, the
.Fl f
flag must be specified to force creation of a new file.  If the
.Fl v
flag is specified,
.Nm chkconfig
displays the new status on the standard output if successful.
If unsuccessful,
.Nm chkconfig
displays an error message and exits with a non-zero return value.
.Sh SEE ALSO
.Xr rc 8
.Sh FILES
.Bl -tag -width /etc/config -compact
.It Pa /etc/config
default location for system configuration files
.El
.Sh BUGS
The
.Fl f
option is bogus, and kept only for SGI compatibility.
.Sh HISTORY
The
.Nm chkconfig
command is modeled after the command of the same name in Silicon
Graphics' IRIX.
.Sh AUTHOR
.Nm chkconfig
was written by Garrett A. Wollman, University of Vermont.


sbin/chkconfig/chkconfig.c
/*
 * Copyright 1993, Garrett A. Wollman.
 * Copyright 1993, University of Vermont and State Agricultural College.
 * All Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE UNIVERSITY AND AUTHOR ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE UNIVERSITY OR AUTHOR BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

const char chkconfig_c_rcsid[] =
  "$Id: chkconfig.c,v 1.1 1993/11/11 05:25:28 wollman Exp $";

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <dirent.h>
#include <errno.h>
#include <string.h>
#include "paths.h"

static int testvalue(const char *);
static int setvalue(const char *, int);
static int printvalues(void);
static void usage(void);
static void die(const char *);
static int is_on(const char *);

const char *whoami;
static const char *configdir = _PATH_CONFIG;
static int sortbystate = 0;
static int force = 0;
static int verbose = 0;

int main(int argc, char **argv) {
  int opt;

  whoami = argv[0];

  while((opt = getopt(argc, argv, "sfd:v")) != EOF) {
    switch(opt) {
    case 's':
      sortbystate = 1;
      break;
    case 'f':
      force = 1;
      break;
    case 'd':
      configdir = optarg;
      break;
    case 'v':
      verbose = 1;
      break;
    case '?':
    default:
      usage();
      return -1;
    }
  }

  if(sortbystate && optind != argc) {
    fprintf(stderr, "%s: -s: too many arguments\n", whoami);
    usage();
    return -1;
  }

  if(force && optind == argc) {
    fprintf(stderr, "%s: -f: too few arguments\n", whoami);
    usage();
    return -1;
  }

  switch(argc - optind) {
  case 0:
    return printvalues();

  case 1:
    return testvalue(argv[optind]);

  case 2:
    return setvalue(argv[optind], is_on(argv[optind + 1]));
  
  default:
    usage();
    return -1;
  }
}

static int is_on(const char *str) {
  if(!str) return 0;
  return (   ((str[0] == 'o') || (str[0] == 'O'))
	  && ((str[1] == 'n') || (str[1] == 'N'))
	  && ((str[2] == '\n')|| (str[2] == '\0')));
}

static void chat(const char *str, int state) {
  if(verbose)
    printf("`%s' is %s.\n", str, state ? "ON" : "OFF");
}

static const char *confname(const char *str) {
  int len = strlen(configdir) + strlen(str) + 2;
  char *rv = malloc(len);

  if(!rv) {
    errno = ENOMEM;
    die("confname: malloc");
  }

  strcpy(rv, configdir);
  strcat(rv, "/");
  strcat(rv, str);
  return rv;
}

static int testvalue(const char *str) {
  FILE *fp;
  char *line;
  const char *fname;
  int rv = 1;			/* NB: shell's convention is opposite C's */

  fname = confname(str);
  fp = fopen(fname, "r");
  if(fp) {
    line = fgetline(fp, (size_t *)0);
    rv = !is_on(line);		/* shell's convention is opposite C's */
    fclose(fp);
  }

  free((void *)fname);		/* cast away `const' to avoid warning */
  chat(str, !rv);		/* tell the user about it if verbose */
  return rv;
}

static int setvalue(const char *str, int state) {
  FILE *fp;
  const char *fname;

  fname = confname(str);
  errno = 0;
  fp = fopen(fname, "r");
  if(!fp && !force) {
    /*
     * Yes, I know this is bogus, but SGI must have had some sort of
     * reason...
     */
    if(errno == ENOENT) {
      fprintf(stderr, 
	      "%s: configuration file does not exist;"
	      " use `-f' flag to create it.\n", whoami);
      return -1;
    } else {
      die(fname);
    }
  }

  if(fp)
    fclose(fp);

  errno = 0;
  fp = fopen(fname, "w");
  if(!fp)
    die(fname);

  fprintf(fp, "%s\n", state ? "on" : "off");

  fclose(fp);

  chat(str, state);
  return 0;
}

/*
 * From here down is the code for listing the value of every option.
 */
struct q {
  struct q *next;
  int state;
  char *name;
};

static struct q *onhead;
static struct q *offhead;	/* only used for sortbystate == 1 */

static void insert(const char *fname) {
  struct q *q;
  struct q **headp;
  int state;

  q = malloc(sizeof *q);
  if(!q) {
    errno = ENOMEM;
    die("insert: malloc");
  }

  q->name = strdup(fname);
  q->state = state = !testvalue(fname);

  if(state || !sortbystate)
    headp = &onhead;
  else
    headp = &offhead;

  while(*headp) {
    if(strcmp(q->name, (**headp).name) >= 0)
      break;			/* found the right spot! */

    headp = &(**headp).next;
  }

  q->next = *headp;
  *headp = q;
}

static int printvalues(void) {
  struct q *temp;
  DIR *dir;
  struct dirent *entry;
  struct stat stab;
  int doneheader = 0;

  verbose = 0;			/* we're already verbose enough */

  errno = 0;
  dir = opendir(configdir);
  if(!dir)
    die(configdir);

  while((entry = readdir(dir))) {
    if(stat(entry->d_name, &stab) < 0) {
      die(entry->d_name);
    }

    if(S_ISREG(stab.st_mode) && !strchr(entry->d_name, '.')) {
      insert(entry->d_name);
    }
  }

  closedir(dir);

  /*
   * Now we're done reading the file names, so we can print them out.
   * Thanks to insert(), everything is already in ASCII order.
   */
#define FORMAT "%15s %-3s\n"

  if(sortbystate) {
    printf("%15s %s\n", "Option", "State");
    printf("%15s %s\n", "======", "=====");
    doneheader = 1;

    while((temp = offhead)) {
      printf(FORMAT, temp->name, "off");
      free(temp->name);
      offhead = temp->next;
      free(temp);
    }
  }

  if(!doneheader) {
    printf("%15s %s\n", "Option", "State");
    printf("%15s %s\n", "======", "=====");
  }

  while((temp = onhead)) {
    printf(FORMAT, temp->name, temp->state ? "on" : "off");
    free(temp->name);
    onhead = temp->next;
    free(temp);
  }

  return 0;
}


static void die(const char *why) {
  fprintf(stderr, "%s: %s: %s\n", whoami, why, strerror(errno));
  exit(-1);
}

static void usage(void) {
  fprintf(stderr, "%s: usage:\n"
	  "%s [ -d configdir ] [ -s ]\n"
	  "%s [ -d configdir ] option\n"
	  "%s [ -d configdir ] [ -f ] option [ on | off ]\n",
	  whoami, whoami, whoami, whoami);
}
